File: app/src/main/java/com/kite/mnemoai/viewmodels/ReciteWordViewModel.java
package com.kite.mnemoai.viewmodels;

import static androidx.lifecycle.SavedStateHandleSupport.createSavedStateHandle;
import static androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY;

import android.os.SystemClock;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.SavedStateHandle;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.viewmodel.ViewModelInitializer;

import com.kite.mnemoai.MainApplication;
import com.kite.mnemoai.data.local.UserSetting;
import com.kite.mnemoai.data.model.DayPlanWord;
import com.kite.mnemoai.data.model.WordExtract;
import com.kite.mnemoai.data.repository.GroupRepository;
import com.kite.mnemoai.data.repository.IRepositoryCallback;
import com.kite.mnemoai.data.repository.StatisticsRepository;
import com.kite.mnemoai.data.repository.UserSettingRepository;
import com.kite.mnemoai.data.repository.WordRepository;
import com.kite.mnemoai.model.WordDetailStatus;
import com.kite.mnemoai.model.WordWithExtractAndDayPlan;
import com.kite.mnemoai.uistate.ReciteWordUIState;
import com.kite.mnemoai.model.WordDetailInfo;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.function.Function;
import java.util.stream.Collectors;

public class ReciteWordViewModel extends ViewModel {
    private WordRepository wordRepository;
    private GroupRepository groupRepository;
    private StatisticsRepository statisticsRepository;
    private UserSettingRepository userSettingRepository;
    private final MediatorLiveData<ReciteWordUIState> _uiState = new MediatorLiveData<>();
    private ReciteWordUIState.ReciteWordStatus reciteWordStatus;
    private List<ReciteWordUIState.ReciteWordItemStatus> reciteWordItemStatuses;
    private List<ReciteWordUIState.ReciteWordItemStatus> reciteWordItemStatusesOrder;
    private List<ReciteStatistics> reciteStatistics = new ArrayList<>();
    private int totalProgress;
    private int currentProgress;
    private boolean shouldAdvance;
    private Random random = new Random();
    private LocalDate date = LocalDate.now();
    private boolean hasSetDailyPlanWord = false;

    public ReciteWordViewModel(WordRepository wordRepository, GroupRepository groupRepository, StatisticsRepository statisticsRepository, UserSettingRepository userSettingRepository, SavedStateHandle savedStateHandle){
        this.wordRepository = wordRepository;
        this.groupRepository = groupRepository;
        this.statisticsRepository = statisticsRepository;
        this.userSettingRepository = userSettingRepository;
        _uiState.addSource(wordRepository.getDailyReciteStatus(), (status) -> {
            switch (status){
                case 0:
                    this.reciteWordStatus = ReciteWordUIState.ReciteWordStatus.no_vocabulary;
                    break;
                case 1:
                    this.reciteWordStatus = ReciteWordUIState.ReciteWordStatus.finish;
                    break;
                case 2:
                    this.reciteWordStatus = ReciteWordUIState.ReciteWordStatus.ok;
                    break;
            }
            updateUIStatus();
        });

        _uiState.addSource(userSettingRepository.getUserSettingLiveData(), userSetting -> {
            wordRepository.setDailyDayPlanWordEntities(userSetting.getNewLearningWordCount());
            hasSetDailyPlanWord = true;
            _uiState.removeSource(userSettingRepository.getUserSettingLiveData());
        });

        _uiState.addSource(wordRepository.getUnfinishPlanWordDetailLiveData(), wordWithExtractAndDayPlans -> {
            wordWithExtractAndDayPlans.forEach(info -> {
                String phonetic = "\\" + info.getWord().getPhonetic() + "\\";
                String translation = info.getWord().getTranslation().replace("\\n", " ");
                info.getWord().setPhonetic(phonetic);
                info.getWord().setTranslation(translation);
            });

            updateOrderData(wordWithExtractAndDayPlans);
            updateUIStatus();
        });

        _uiState.addSource(statisticsRepository.getAllPlanCountByDate(date.toString()), (allPlanCount) -> {
            this.totalProgress = allPlanCount;
            updateUIStatus();
        });

        _uiState.addSource(statisticsRepository.getFinishedPlanCountByDate(date.toString()), (finishedPlanCount) -> {
            this.currentProgress = finishedPlanCount;
            updateUIStatus();
        });
    }

    private void updateOrderData(List<WordWithExtractAndDayPlan> newPlans) {
        if (this.reciteWordItemStatusesOrder == null) {
            // 首次初始化：直接创建新的状态列表
            this.reciteWordItemStatusesOrder = newPlans.stream()
                    .map(plan -> new ReciteWordUIState.ReciteWordItemStatus(plan, WordDetailStatus.hide, false))
                    .collect(Collectors.toList());
        } else {

            Map<Long, WordWithExtractAndDayPlan> newPlanMap = newPlans.stream()
                    .collect(Collectors.toMap(
                            plan -> plan.getWord().getId(),
                            Function.identity(),
                            (existing, replacement) -> replacement
                    ));

            // 1. 新列表不存在，旧列表存在
            Iterator<ReciteWordUIState.ReciteWordItemStatus> iterator = this.reciteWordItemStatusesOrder.iterator();
            while (iterator.hasNext()){
                ReciteWordUIState.ReciteWordItemStatus reciteWordItemStatus = iterator.next();
                long wordId = reciteWordItemStatus.getWordWithExtractAndDayPlan().getWord().getId();
                WordWithExtractAndDayPlan wordWithExtractAndDayPlan = newPlanMap.get(wordId);
                if(wordWithExtractAndDayPlan == null){
                    iterator.remove();
                }
            }

            Map<Long, ReciteWordUIState.ReciteWordItemStatus> oldStatusMap =
                    this.reciteWordItemStatusesOrder.stream()
                            .collect(Collectors.toMap(
                                    status -> status.getWordWithExtractAndDayPlan().getWord().getId(),
                                    Function.identity()
                            ));

            // 2. 新列表存在，旧序列不存在
            for(WordWithExtractAndDayPlan newPlan: newPlans){
                long worldId = newPlan.getWord().getId();
                ReciteWordUIState.ReciteWordItemStatus oldStatus = oldStatusMap.get(worldId);
                if(oldStatus == null){
                    int orderSize = this.reciteWordItemStatusesOrder.size();
                    if(orderSize == 0){
                        this.reciteWordItemStatusesOrder.add(0, new ReciteWordUIState.ReciteWordItemStatus(newPlan, WordDetailStatus.hide, false));
                    }else{
                        int insertIndex = random.nextInt(orderSize) + 1;
                        this.reciteWordItemStatusesOrder.add(insertIndex, new ReciteWordUIState.ReciteWordItemStatus(newPlan, WordDetailStatus.hide, false));
                    }
                }else{
                    //赋新值
                    oldStatus.setWordWithExtractAndDayPlan(newPlan);
                }
            }
        }
    }

    public void setDailyDayPlanWordEntities(){
        userSettingRepository.getUserSetting(new IRepositoryCallback<UserSetting>() {
            @Override
            public void onComplete(UserSetting userSetting) {
                wordRepository.setDailyDayPlanWordEntities(userSetting.getNewLearningWordCount());
            }

            @Override
            public void onError(Throwable t) {

            }
        });
    }

    private void updateUIStatus(){
        _uiState.setValue(new ReciteWordUIState(reciteWordStatus, reciteWordItemStatuses, reciteWordItemStatusesOrder, totalProgress, currentProgress, shouldAdvance));
    }

    public void fetchWordExtract(WordWithExtractAndDayPlan wordWithExtractAndDayPlan, IRepositoryCallback<WordExtract> callback){
        wordRepository.fetchWordExtract(wordWithExtractAndDayPlan, callback);
    }

    public LiveData<ReciteWordUIState> getUiState() {
        return _uiState;
    }

    public List<ReciteStatistics> getReciteStatistics() {
        return reciteStatistics;
    }

    public void showTranslation(){
        if(this.reciteWordItemStatusesOrder == null || this.reciteWordItemStatusesOrder.isEmpty()) return;
        ReciteWordUIState.ReciteWordItemStatus showed = this.reciteWordItemStatusesOrder.get(0);
        showed.setShowTranslation(true);
        if(showed.getWordWithExtractAndDayPlan().getWordExtract() == null){
            showed.setWordDetailStatus(WordDetailStatus.loading);
            fetchWordExtract(showed.getWordWithExtractAndDayPlan(), new IRepositoryCallback<WordExtract>() {
                @Override
                public void onComplete(WordExtract wordExtract) {
                    if(reciteWordItemStatusesOrder.get(0) == showed) showed.setWordDetailStatus(WordDetailStatus.show);
                    else showed.setWordDetailStatus(WordDetailStatus.hide);
                }

                @Override
                public void onError(Throwable t) {
                    if(reciteWordItemStatusesOrder.get(0) == showed) showed.setWordDetailStatus(WordDetailStatus.error);
                    else showed.setWordDetailStatus(WordDetailStatus.hide);
                }
            });
        }else {
            showed.setWordDetailStatus(WordDetailStatus.show);
        }
        updateUIStatus();
    }
    public void rememberWord(){
        if(this.reciteWordItemStatusesOrder == null || this.reciteWordItemStatusesOrder.isEmpty()) return;
        ReciteWordUIState.ReciteWordItemStatus remembered = this.reciteWordItemStatusesOrder.get(0);
        this.reciteWordItemStatusesOrder.remove(remembered);
        ReciteStatistics rememberedReciteStatistics = reciteStatistics.stream()
                .filter(rs -> rs.getWordId() == remembered.getWordWithExtractAndDayPlan().getWord().getId()).findFirst().get();
        updateReciteStatistics(remembered.getWordWithExtractAndDayPlan().getWord().getId(), 2);
        statisticsRepository.rememberWord(rememberedReciteStatistics);
        this.shouldAdvance = true;
        updateUIStatus();
    }

    public void blurWord() {
        if(this.reciteWordItemStatusesOrder == null || this.reciteWordItemStatusesOrder.isEmpty()) return;
        ReciteWordUIState.ReciteWordItemStatus blured = this.reciteWordItemStatusesOrder.get(0);
        resetShowStatus(blured);

        this.reciteWordItemStatusesOrder.remove(blured);
        int size = this.reciteWordItemStatusesOrder.size();

        // 如果列表为空，直接添加并返回（虽然下方也会处理，但更清晰）
        if (size == 0) {
            this.reciteWordItemStatusesOrder.add(blured);
            updateUIStatus();
            return;
        }

        // 计算 30%~60% 区间的整数索引
        int minPos = (int) (0.6 * size);
        int maxPos = (int) (size);

        // 保证不越界
        minPos = Math.max(0, minPos);
        maxPos = Math.min(size, maxPos);
        // 理论上 minPos <= maxPos，但防止浮点误差
        if (minPos > maxPos) minPos = maxPos;

        // 随机位置（包含两端）
        int position = minPos + random.nextInt(maxPos - minPos + 1);
        this.reciteWordItemStatusesOrder.add(position, blured);
        this.shouldAdvance = true;
        updateReciteStatistics(blured.getWordWithExtractAndDayPlan().getWord().getId(), 1);
        updateUIStatus();
    }

    public void forgetWord() {
        if(this.reciteWordItemStatusesOrder == null || this.reciteWordItemStatusesOrder.isEmpty()) return;
        ReciteWordUIState.ReciteWordItemStatus forgot = this.reciteWordItemStatusesOrder.get(0);
        resetShowStatus(forgot);

        this.reciteWordItemStatusesOrder.remove(forgot);
        int size = this.reciteWordItemStatusesOrder.size();

        // 如果列表为空，直接添加并返回（虽然下方也会处理，但更清晰）
        if (size == 0) {
            this.reciteWordItemStatusesOrder.add(forgot);
            updateUIStatus();
            return;
        }

        // 计算 60%~100% 区间的整数索引
        int minPos = (int) (0.3 * size);
        int maxPos = (int) (0.6 * size);
        // 保证不越界
        minPos = Math.max(0, minPos);
        maxPos = Math.min(size, maxPos);
        // 理论上 minPos <= maxPos，但防止浮点误差
        if (minPos > maxPos) minPos = maxPos;

        // 随机位置（包含两端）
        int position = minPos + random.nextInt(maxPos - minPos + 1);
        this.reciteWordItemStatusesOrder.add(position, forgot);
        this.shouldAdvance = true;
        updateReciteStatistics(forgot.getWordWithExtractAndDayPlan().getWord().getId(), 0);
        updateUIStatus();
    }

    public boolean isShouldAdvance() {
        return shouldAdvance;
    }

    public void setShouldAdvance(boolean shouldAdvance) {
        this.shouldAdvance = shouldAdvance;
    }

    public void addReciteStatistics(){
        if(this.reciteWordItemStatusesOrder == null || this.reciteWordItemStatusesOrder.isEmpty())return;
        this.reciteStatistics.add(new ReciteStatistics(this.reciteWordItemStatusesOrder.get(0).getWordWithExtractAndDayPlan().getWord().getId(), SystemClock.elapsedRealtime()));
    }

    public void updateReciteStatistics(long wordId, int type){
        //type修改类型， 0 - 模糊， 1 - 忘记， 2 - 记住
        for (int i = 0; i< this.reciteStatistics.size(); i++){
            ReciteStatistics currentReciteStatistics = this.reciteStatistics.get(i);
            if(currentReciteStatistics.getWordId() == wordId){
                if(type == 0){
                    currentReciteStatistics.addBlurCount();
                    currentReciteStatistics.addLearningTime(SystemClock.elapsedRealtime() - currentReciteStatistics.getStartLearningTime());
                } else if (type == 1) {
                    currentReciteStatistics.addForgetCount();
                    currentReciteStatistics.addLearningTime(SystemClock.elapsedRealtime() - currentReciteStatistics.getStartLearningTime());
                } else if (type == 2) {
                    currentReciteStatistics.addLearningTime(SystemClock.elapsedRealtime() - currentReciteStatistics.getStartLearningTime());
                }
            }
        }
    }

    public void resetShowStatus(ReciteWordUIState.ReciteWordItemStatus reciteWordItemStatus){
        if(reciteWordItemStatus == null) return;
        reciteWordItemStatus.setShowTranslation(false);
        reciteWordItemStatus.setWordDetailStatus(WordDetailStatus.hide);
    }

    public static final ViewModelInitializer<ReciteWordViewModel> initializer = new ViewModelInitializer<>(
            ReciteWordViewModel.class,
            creationExtras -> {
                MainApplication app = (MainApplication) creationExtras.get(APPLICATION_KEY);
                assert app != null;
                SavedStateHandle savedStateHandle = createSavedStateHandle(creationExtras);
                return new ReciteWordViewModel(app.getWordRepository(), app.getGroupRepository(), app.getStatisticsRepository(), app.getUserSettingRepository(), savedStateHandle);
            }
    );

    public static class ReciteStatistics{
        private long wordId;
        private int blurCount;
        private int forgetCount;
        private long startLearningTime;
        private long learningTime;

        public ReciteStatistics(long wordId, long startLearningTime) {
            this.wordId = wordId;
            this.startLearningTime = startLearningTime;
        }

        public long getWordId() {
            return wordId;
        }

        public long getStartLearningTime() {
            return startLearningTime;
        }

        public void addLearningTime(long time){
            this.learningTime = this.learningTime + time;
        }

        public void addBlurCount(){
            blurCount = blurCount + 1;
        }
        public void addForgetCount(){
            forgetCount = forgetCount + 1;
        }

        public int getBlurCount() {
            return blurCount;
        }

        public int getForgetCount() {
            return forgetCount;
        }

        public long getLearningTime() {
            return learningTime;
        }
    }
}

File: feature/reciteword/src/main/java/com/kite/mnemoai/reciteword/ReciteWordViewModel.kt
package com.kite.mnemoai.reciteword

import android.net.Uri
import android.os.SystemClock
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.kite.domain.RememberWordUseCase
import com.kite.domain.SetDailyReciteWordUseCase
import com.kite.mnemoai.model.Result
import com.kite.mnemoai.model.recite.PronounceType
import com.kite.mnemoai.model.repository.AiMnemonicRepository
import com.kite.mnemoai.model.repository.ReciteStatistics
import com.kite.mnemoai.model.repository.StatisticsRepository
import com.kite.mnemoai.model.repository.UserSettingRepository
import com.kite.mnemoai.model.repository.WordRepository
import com.kite.mnemoai.model.word.WordDetail
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.util.Random
import javax.inject.Inject
import kotlin.math.max
import kotlin.math.min

@OptIn(ExperimentalCoroutinesApi::class)
@HiltViewModel
class ReciteWordViewModel @Inject constructor(
    private val wordRepository: WordRepository,
    private val statisticsRepository: StatisticsRepository,
    private val userSettingRepository: UserSettingRepository,
    private val aiMnemonicRepository: AiMnemonicRepository,
    private val rememberWordUseCase: RememberWordUseCase,
    private val setDailyReciteWordUseCase: SetDailyReciteWordUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow<ReciteWordUIState?>(null)
    val uiState: StateFlow<ReciteWordUIState?> = _uiState.asStateFlow()
    private var reciteStage: ReciteStage? = null
    private var reciteWordDetailInfoItemUIState: MutableList<WordDetail>? = null
    private var reciteStatistics: MutableList<ReciteStatistics> = mutableListOf()
    private var totalProgress = 0
    private var currentProgress = 0
    private val _showNext = MutableSharedFlow<Unit>(extraBufferCapacity = 1)
    val showNext: SharedFlow<Unit> = _showNext.asSharedFlow()
    private var isShowTranslation = false
    private var isShowDetail = false
    private val _aiMnemonicLoadingState = MutableSharedFlow<Result<String>?>(extraBufferCapacity = 1)
    val aiMnemonicLoadingState get() = _aiMnemonicLoadingState.asSharedFlow()
    private var revision = 0L
    private val random = Random()
    private val today = MutableStateFlow(LocalDate.now())
    private var apiKey: String? = null
    private var autoPronounce = true
    private var pronounceType: PronounceType = PronounceType.USA


    init {
        viewModelScope.launch {
            today.flatMapLatest { date ->
                wordRepository.observeDailyReciteStatus(date.toString())
            }.collect { result ->
                if (result is Result.Success) {
                    reciteStage = when (result.data) {
                        0 -> ReciteStage.NO_VOCABULARY
                        1 -> ReciteStage.FINISH
                        2 -> ReciteStage.IN_PROGRESS
                        3 -> ReciteStage.NOT_STARTED
                        else -> null
                    }
                    updateUIStatus()
                }
            }
        }
        viewModelScope.launch {
            userSettingRepository.userSetting.collect { result ->
                apiKey = result.deepseekSettings.apiKey
                autoPronounce = result.autoPronounce
                pronounceType = result.pronounceType
                updateUIStatus()
            }
        }
        viewModelScope.launch {
            today.flatMapLatest { date ->
                wordRepository.observeUnfinishedWordDetailInfos(date.toString())
            }.collect { result ->
                if (result is Result.Success) {
                    val formatted = result.data.map { detail -> formatWordDetail(detail) }
                    updateOrderData(formatted.toMutableList())
                    updateUIStatus()
                }
            }
        }
        viewModelScope.launch {
            today.flatMapLatest { date ->
                statisticsRepository.observeAllPlanCountByDate(date.toString())
            }.collect { result ->
                if (result is Result.Success) {
                    totalProgress = result.data
                    updateUIStatus()
                }
            }
        }
        viewModelScope.launch {
            today.flatMapLatest { date ->
                statisticsRepository.observeFinishedPlanCountByDate(date.toString())
            }.collect { result ->
                if (result is Result.Success) {
                    currentProgress = result.data
                    updateUIStatus()
                }
            }
        }
    }

    private fun formatWordDetail(detail: WordDetail): WordDetail {
        val updateDayPlanWords = detail.dayPlanWords.filter { it.date != today.value.toString()}
        val updatedWord = detail.word.copy(phonetic = "\\${detail.word.phonetic}\\")
        val sortedTranslations = detail.translations.sorted()
        val updatedTranslations = sortedTranslations.map { translation ->
            translation.copy(
                meanings = translation.meanings.map { meaning ->
                    meaning.copy(meaning = "${meaning.meaning};")
                }
            )
        }
        return detail.copy(word = updatedWord, translations = updatedTranslations, dayPlanWords = updateDayPlanWords)
    }

    private fun updateOrderData(newPlans: MutableList<WordDetail>) {
        if (this.reciteWordDetailInfoItemUIState == null) {
            this.reciteWordDetailInfoItemUIState = newPlans
        } else {
            val newPlanMap = newPlans.associateBy { it.word.id }

            // 1. 旧列表中存在、新列表中不存在 → 移除
            val iterator = this.reciteWordDetailInfoItemUIState!!.iterator()
            while (iterator.hasNext()) {
                val wordDetailInfo = iterator.next()
                if (newPlanMap[wordDetailInfo.word.id] == null) {
                    iterator.remove()
                    _showNext.tryEmit(Unit)
                    resetShowState()
                }
            }

            val oldWordDetailInfoMap =
                this.reciteWordDetailInfoItemUIState!!.associateBy { it.word.id }

            // 2. 新列表中存在、旧列表中不存在 → 插入；存在 → 更新字段
            for (newPlan in newPlans) {
                val wordId = newPlan.word.id
                val oldStatus = oldWordDetailInfoMap[wordId]
                if (oldStatus == null) {
                    val orderSize = this.reciteWordDetailInfoItemUIState!!.size
                    if (orderSize == 0) {
                        this.reciteWordDetailInfoItemUIState!!.add(newPlan)
                    } else {
                        val insertIndex = random.nextInt(orderSize) + 1
                        this.reciteWordDetailInfoItemUIState!!.add(insertIndex, newPlan)
                    }
                    _showNext.tryEmit(Unit)
                    resetShowState()
                } else {
                    val index = this.reciteWordDetailInfoItemUIState!!.indexOf(oldStatus)
                    if (index >= 0) {
                        this.reciteWordDetailInfoItemUIState!![index] = oldStatus.copy(
                            word = newPlan.word,
                            extract = newPlan.extract,
                            dayPlanWords = newPlan.dayPlanWords
                        )
                    }
                }
            }
        }
    }

    fun setDailyDayPlanWordEntities() {
        viewModelScope.launch {
            setDailyReciteWordUseCase()
        }
    }

    fun onResume() {
        if (LocalDate.now() == today.value) return
        stopReciteStatistics()
        today.value = LocalDate.now()
        reciteStage = null
        reciteWordDetailInfoItemUIState = null
        reciteStatistics.clear()
        totalProgress = 0
        currentProgress = 0
        isShowTranslation = false
        isShowDetail = false
        setDailyDayPlanWordEntities()
    }

    private fun updateUIStatus() {
        revision++
        _uiState.value = ReciteWordUIState(
            reciteStage,
            reciteWordDetailInfoItemUIState,
            totalProgress,
            currentProgress,
            isShowTranslation,
            isShowDetail,
            revision,
            autoPronounce = autoPronounce,
            pronounceType = pronounceType
        )
    }

    fun showAll() {
        this.isShowTranslation = true
        this.isShowDetail = true
        updateUIStatus()
    }

    fun resetShowState() {
        this.isShowTranslation = false
        this.isShowDetail = false
        updateUIStatus()
    }

    fun fetchWordExtract() {
        if (this.reciteWordDetailInfoItemUIState.isNullOrEmpty()) return
        val currentWord = this.reciteWordDetailInfoItemUIState!![0]
        _aiMnemonicLoadingState.tryEmit(Result.Loading)
        viewModelScope.launch {
            _aiMnemonicLoadingState.tryEmit(aiMnemonicRepository.generateWordExtract(listOf(currentWord)))
        }
    }

    fun rememberWord() {
        if (this.reciteWordDetailInfoItemUIState.isNullOrEmpty()) return
        val remembered = this.reciteWordDetailInfoItemUIState!![0]
        resetShowState()
        this.reciteWordDetailInfoItemUIState!!.remove(remembered)
        ensureReciteStatistics(remembered.word.id)
        val rememberedReciteStatistics = reciteStatistics.first { rs -> rs.wordId == remembered.word.id }
        updateReciteStatistics(remembered.word.id, 2)
        viewModelScope.launch {
            rememberWordUseCase(rememberedReciteStatistics)
        }
        _showNext.tryEmit(Unit)
        updateUIStatus()
    }

    fun blurWord() {
        if (this.reciteWordDetailInfoItemUIState.isNullOrEmpty()) return
        val blured = this.reciteWordDetailInfoItemUIState!![0]
        resetShowState()

        this.reciteWordDetailInfoItemUIState!!.remove(blured)
        val size = this.reciteWordDetailInfoItemUIState!!.size

        if (size == 0) {
            this.reciteWordDetailInfoItemUIState!!.add(blured)
            updateUIStatus()
            return
        }

        var minPos = (0.6 * size).toInt()
        var maxPos = size
        minPos = max(0, minPos)
        maxPos = min(size, maxPos)
        if (minPos > maxPos) minPos = maxPos

        val position = minPos + random.nextInt(maxPos - minPos + 1)
        this.reciteWordDetailInfoItemUIState!!.add(position, blured)
        _showNext.tryEmit(Unit)
        ensureReciteStatistics(blured.word.id)
        updateReciteStatistics(blured.word.id, 1)
        updateUIStatus()
    }

    fun forgetWord() {
        if (this.reciteWordDetailInfoItemUIState.isNullOrEmpty()) return
        val forgot = this.reciteWordDetailInfoItemUIState!![0]
        resetShowState()

        this.reciteWordDetailInfoItemUIState!!.remove(forgot)
        val size = this.reciteWordDetailInfoItemUIState!!.size

        if (size == 0) {
            this.reciteWordDetailInfoItemUIState!!.add(forgot)
            updateUIStatus()
            return
        }

        var minPos = (0.3 * size).toInt()
        var maxPos = (0.6 * size).toInt()
        minPos = max(0, minPos)
        maxPos = min(size, maxPos)
        if (minPos > maxPos) minPos = maxPos

        val position = minPos + random.nextInt(maxPos - minPos + 1)
        this.reciteWordDetailInfoItemUIState!!.add(position, forgot)
        _showNext.tryEmit(Unit)
        ensureReciteStatistics(forgot.word.id)
        updateReciteStatistics(forgot.word.id, 0)
        updateUIStatus()
    }

    /**
     * 确保指定单词的学习统计项已建立。
     * 修复竞态：进程重建后 Room 数据流晚于 onStart 到达时，
     * startReciteStatistics 会因列表为空提前返回，导致 reciteStatistics 缺失当前词统计项，
     * 进而使 rememberWord 的 first{} 抛 NoSuchElementException、blur/forget 的计数丢失。
     */
    private fun ensureReciteStatistics(wordId: Long) {
        if (reciteStatistics.none { it.wordId == wordId }) {
            reciteStatistics.add(
                ReciteStatistics(
                    wordId = wordId,
                    startLearningTime = SystemClock.elapsedRealtime()
                )
            )
        }
    }

    fun startReciteStatistics() {
        if (this.reciteWordDetailInfoItemUIState.isNullOrEmpty()) return
        val currentWordId = this.reciteWordDetailInfoItemUIState!![0].word.id
        val existing = reciteStatistics.firstOrNull { rs -> rs.wordId == currentWordId }
        if (existing != null) {
            existing.startLearningTime = SystemClock.elapsedRealtime()
        } else {
            reciteStatistics.add(
                ReciteStatistics(
                    wordId = currentWordId,
                    startLearningTime = SystemClock.elapsedRealtime()
                )
            )
        }
    }

    fun getMediaItemURI(): String {
        if (this.reciteWordDetailInfoItemUIState.isNullOrEmpty()) return ""
        val currentWord = this.reciteWordDetailInfoItemUIState!![0].word.word
        val type = when (pronounceType) {
            PronounceType.USA -> 2
            PronounceType.UK -> 1
        }
        return "https://dict.youdao.com/dictvoice?audio=${Uri.encode(currentWord)}&type=$type"
    }

    fun isAutoPronounceEnabled(): Boolean = autoPronounce

    fun currentPronounceType(): PronounceType = pronounceType
    fun updateReciteStatistics(wordId: Long, type: Int) {
        for (currentReciteStatistics in this.reciteStatistics) {
            if (currentReciteStatistics.wordId == wordId) {
                when (type) {
                    0 -> {
                        currentReciteStatistics.addForgetCount()
                        currentReciteStatistics.addLearningTime(SystemClock.elapsedRealtime() - currentReciteStatistics.startLearningTime)
                    }
                    1 -> {
                        currentReciteStatistics.addBlurCount()
                        currentReciteStatistics.addLearningTime(SystemClock.elapsedRealtime() - currentReciteStatistics.startLearningTime)
                    }
                    2, 3 -> {
                        currentReciteStatistics.addLearningTime(SystemClock.elapsedRealtime() - currentReciteStatistics.startLearningTime)
                    }
                }
            }
        }
    }

    fun stopReciteStatistics() {
        if (this.reciteWordDetailInfoItemUIState.isNullOrEmpty()) return
        val currentWordId = this.reciteWordDetailInfoItemUIState!![0].word.id
        updateReciteStatistics(currentWordId, 3)
    }

    fun saveEnableAutoPronounce(enableAutoPronounce: Boolean){
        viewModelScope.launch {
            userSettingRepository.setAutoPronounce(enableAutoPronounce)
        }
    }

    fun savePronounceType(pronounceType: PronounceType){
        viewModelScope.launch {
            userSettingRepository.setPronounceType(pronounceType)
        }
    }
}
